mlir/lib/Bindings/Python/Rewrite.cpp

```diff
@@ -36,6 +41,43 @@ class PyPDLPatternModule {
   }
   MlirPDLPatternModule get() { return module; }
 
+  static nb::object fromPDLValue(MlirPDLValue value) {
+    if (mlirPDLValueIsValue(value)) {
+      return nb::cast(mlirPDLValueAsValue(value));
+    }
+    if (mlirPDLValueIsOperation(value)) {
+      return nb::cast(mlirPDLValueAsOperation(value));
+    }
+    if (mlirPDLValueIsAttribute(value)) {
+      return nb::cast(mlirPDLValueAsAttribute(value));
+    }
+    if (mlirPDLValueIsType(value)) {
+      return nb::cast(mlirPDLValueAsType(value));
+    }
+
+    throw std::runtime_error("unsupported PDL value type");
+  }
+
+  void registerRewriteFunction(const std::string &name,
+                               const nb::callable &fn) {
+    mlirPDLPatternModuleRegisterRewriteFunction(
+        get(), mlirStringRefCreate(name.data(), name.size()),
+        [](MlirPatternRewriter rewriter, MlirPDLResultList results,
+           size_t nValues, MlirPDLValue *values,
+           void *userData) -> MlirLogicalResult {
+          auto f = nb::handle(static_cast<PyObject *>(userData));
+          std::vector<nb::object> args;
+          args.reserve(nValues);
+          for (size_t i = 0; i < nValues; ++i) {
+            args.push_back(fromPDLValue(values[i]));
+          }
+          return nb::cast<bool>(f(rewriter, results, args))
+                     ? mlirLogicalResultSuccess()
+                     : mlirLogicalResultFailure();
```

Comment:
I know it's weird but the convetion is that `False` means success and `True` means failure. That way actually an `f` that doesn't have an explicit return value (i.e., returns `None`) will succeed.

Reply:
Looks a little weird but reasonable. I'll change it.

Reply:
I think it's like shell handling of exit codes works (exit 0 means success).

Reply:
Done in e9845d639f6243d6133cd79d4a861563dc6784ca.